# Введение в типы данных

При обработке информации в ЭВМ возникает два вопроса: где хранить информацию и как её обрабатывать. Вопрос "как обрабатывать" -- это вопрос алгоритмов, а "где хранить" -- структур данных. Каждый язык обладает рядом встроенных структур. Типы данных являются их подмножеством.

# Иерархия типов данных в Python

### Числовые типы:
* Целые (Integers, Booleans)
* Нецелые (Floats, Complex, Decimals, Fractions)

### Последовательности:
* Изменяемые (Lists)
* Неизменяемые (Tuples, Strings)

### Множества:
* Изменяемые (Sets)
* Неизменяемые (Frozen sets)

### Словари (Dictionaries)

### Вызываемые объекты (Callables):
* Функции определенные пользователем (user-defined functions)
* Встроенные функции (built-in functions) (`len()`, `open()`)
* Генераторы (generators)
* Классы (classes)
* Методы класса (instance methods)
* Встроенные методы (built-in methods) (`list.append()`)
* Class instances (`.__call__()`)

### Singeltons:
* None
* NotImplemented
* Ellipsis

Область памяти, в которой хранится некоторое значение, называется **объектом**. Например:
```
x = 1
```
`x` -- **переменная**, а значение 1, которое располагается где-то в памяти является объектом.

**Переменная** -- ссылка на объект; некоторое имя, по которому этот объект может быть идентифицирован. Память представляет собой последовательность ячеек.

In [1]:
x = 1
print(x) # функция вывода; выводит значение переменной

1


Номер ячейки, на которую ссылается переменная можно узнать при помощи функции `id`:

In [2]:
print(id(x))

2592883867888


`=` - операция присваивания. Она работает следующим образом: вычисляется значение выражения, которое находится справа от знака присваивания. В результате этого получается некоторый объект. В дальнейшем имя переменной слева связывается с объектом. Когда выполняется присваивание:

In [3]:
x = 46 + 79
print(x)

125


создаётся ссылка `x`, которая указывает на объект, хранящий в себе значение `46 + 79 = 125`. Объект с таким значением создаётся в памяти, следовательно, занимает место. На данный момент мы имеем одну ссылку на объект.

Допустим, что дальше была создана ещё одна ссылка на тот же объект:

In [4]:
y = x # Переменная y теперь ссылается туда же, куда и x. Они обе знают, где располагается объект.

Можно легко убедиться, что переменные ссылаются на один и тот же объект:

In [5]:
print(id(x), id(y))

2592883871856 2592883871856


Количество ссылок увеличивается на 1. Интерпретатору известно, сколько ссылок указывает на объект и если количество ссылок становится равным нулю, объект уничтожается с целью освободить память.

**Интерпретатор** -- программа, выполняющая интерпретацию. **Интерпрета́ция** — построчный анализ, обработка и выполнение исходного кода программы или запроса.


Для дальнейшего повествования нам потребуется разбор функции вывода `print`. Рассмотрим способы вывода переменных в языке программирования Python:

In [6]:
x = 100

print(x)                 # вывод значения переменной x
print(42)                # вывод значения 42
print("Hello, world")    # вывод сообщения "Hello, world"
print('Hi!')             # вывод сообщения "Hi!"
# "Hello, world" -- строка. Строкой называется последовательность символов, заключенных в двойные или одинарные кавычки.

print(42, x * 10, 1, "Hi!")
print("Значение x =", x) # вывод сообщения, за которым последует значение x

# После каждого вызова функции print происходит перенос на новую строку

100
42
Hello, world
Hi!
42 1000 1 Hi!
Значение x = 100


Последний вывод показанный выше неудобен, когда количество переменных становится велико. Для решения этой задачи в Python добавили f-строки. Строка называется f-строкой, так как перед двойными кавычками (которые являются признаком строки) ставится символ f. F-строки позволяют легко комбинировать вывод переменных и строк:

In [7]:
x = 10
print(f"Значение x = x")   # если нет переменных в {}, то будет выведена как простая строка
print(f"Значение x = {x}") # вместо {x} будет подставлен объект, на который указывает x
print(f"Значение 2x = {x * 2}")

Значение x = x
Значение x = 10
Значение 2x = 20


In [8]:
name = 'Антон'
age = 250
print(f'Его имя {name}. Ему {age} лет')

Его имя Антон. Ему 250 лет


Можно использовать метод `.format` для вывода значений переменных внутри строки.

In [9]:
name = "Виктор"
age = 10

# На место первых {} подставится name, на место вторых {} age
print('Его зовут {}. Ему {} лет'.format(name, age))

Его зовут Виктор. Ему 10 лет


In [10]:
# В этом варианте функции вместо {0} подставится значение переменной name,
# вместо {1} -- значение возраста.
print('Его имя {0}. Просто {0}. Ему {1} лет'.format(name, age))

Его имя Виктор. Просто Виктор. Ему 10 лет


Возвращаемся к рассказу о переменных. Для того, чтобы узнать, сколько ссылок указывает на объект, можно воспользоваться функцией
```sys.getrefcount(x)```
(примечание: возвращаемое количество обычно больше, чем можно ожидать)

In [11]:
import sys # Импортирование библиотеки. Это необходимо, чтобы иметь доступ к функциям sys.

v1 = 523424
print(f"Количество ссылок на объект id = {id(v1)} равняется {sys.getrefcount(v1)}")
v2 = v1  # количество ссылок должно увеличиться на 1
print(f"Количество ссылок на объект id = {id(v2)} равняется {sys.getrefcount(v2)}")
v2 += 1  # количество ссылок на значение 523424 должно уменьшиться на 1
print(f"Количество ссылок на объект id = {id(v1)} равняется {sys.getrefcount(v1)}")

Количество ссылок на объект id = 2592961541584 равняется 3
Количество ссылок на объект id = 2592961541584 равняется 4
Количество ссылок на объект id = 2592961541584 равняется 3


В отличии от языков С / С++ и прочих, тип переменной не связан непосредственно с идентификатором. Когда пишется
```
x = 10
```
Из этого не следует, что `x` -- переменная целочисленного типа. Это значит, что `x` -- ссылка, на целочисленное значение. Она может в будущем указывать на объект другого типа:
```
x = 'hello, world!'
```
Вызывая функцию `type(x)` мы получаем не тип переменной `x`, а тип объекта, на который указывает ссылка.

In [12]:
x = 10
print(type(x))

x = 'Hello world'
print(type(x))

<class 'int'>
<class 'str'>


Когда вычисляется значение выражения, получается некоторый результат. Этому результату должен соответствовать объект. В примере ниже можно убедиться, что адрес расположения объекта, полученного в результате сложения совпадает с адресом переменной, которая хранит такое же значение:

In [13]:
x = 10
print(id(x))
x += 5  # то же самое что и x = x + 5
print(id(x), id(15))

2592883868176
2592883868336 2592883868336


In [14]:
# Функции являются объектами. Они тоже имеют свой адрес в памяти:
print(id(id))

2592884494224
